tc_aws/aws/bucket.py

```diff
@@ -44,24 +46,22 @@ def __init__(self, bucket, region, endpoint, max_retry=None):
                 )
             )
 
-        if self._client is None:
-            self._client = aiobotocore.get_session().create_client(
-                's3',
-                region_name=region,
-                endpoint_url=endpoint,
-                config=config
-            )
+        self.region_name = region
+        self.endpoint_url = endpoint
+        self.session = aiobotocore.session.get_session()
 
     async def exists(self, path):
         """
         Checks if an object exists at a given path
         :param string path: Path or 'key' to retrieve AWS object
         """
         try:
-            await self._client.head_object(
-                Bucket=self._bucket,
-                Key=self._clean_key(path),
-            )
+            async with self.session.create_client('s3', region_name=self.region_name,
```

Comment:
Did you test this with a load-intensive process? We had performance issues in the past without the singleton client

Reply:
I did some tests using `hey` client in that epoch, but I didn't any bad behavior (but didn't  keep the results :( )

I remember that the client was a corutine itself ... but I'll run some tests and paste the results here to help with the analysis.

Originally, I was worried about opening the payload at https://github.com/thumbor-community/aws/pull/157/files#diff-8c5f6e09db7784ddba2fc0a87e8c9e5436275868ae07088bc1f5a1c888c45224R74-R81, but that hint about the client is warm as well.

I'll brb soon with the hey results.

Reply:
Is there any performance tests guidelines or number to make a comparison?

Here is the information about the tests that I made.

In my case, my CPU is the following:

```
processor	: 23
vendor_id	: AuthenticAMD
cpu family	: 23
model		: 113
model name	: AMD Ryzen 9 3900X 12-Core Processor
stepping	: 0
microcode	: 0x8701021
cpu MHz		: 2456.247
cache size	: 512 KB
physical id	: 0
siblings	: 24
core id		: 14
cpu cores	: 12
apicid		: 29
initial apicid	: 29
fpu		: yes
fpu_exception	: yes
cpuid level	: 16
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sme sev sev_es
bugs		: sysret_ss_attrs spectre_v1 spectre_v2 spec_store_bypass
bogomips	: 7585.13
TLB size	: 3072 4K pages
clflush size	: 64
cache_alignment	: 64
address sizes	: 43 bits physical, 48 bits virtual
power management: ts ttp tm hwpstate cpb eff_freq_ro [13] [14]
```

I ran the tests using docker on a Linux Machine in all cases. Each test was run three times and in my scenario, I have an S3 bucket served through minio and a redis cache:

```
version: '3'

volumes:
  redis-data:
  s3-data:
  thumbor-data:
  thumbor-logs:

services:
  s3:
    image: gcavalcante8808/minio-dev:latest
    environment:
      MINIO_ROOT_USER: minio
      MINIO_ROOT_PASSWORD: minio123
      MINIO_INITIAL_BUCKET: default
      MINIO_INITIAL_BUCKET_PERMISSION: none
    volumes:
      - s3-data:/data

  thumbor:
    image: apsl/thumbor:latest
    volumes:
      - thumbor-data:/data
      - thumbor-logs:/logs
    environment:
      - DETECTORS=['thumbor.detectors.queued_detector.queued_complete_detector']
      - STORAGE=thumbor.storages.mixed_storage
      - REDIS_STORAGE_SERVER_HOST=redis
      - REDIS_STORAGE_SERVER_PORT=6379
      - REDIS_STORAGE_SERVER_DB=0
      - REDIS_QUEUE_SERVER_HOST=redis
      - REDIS_QUEUE_SERVER_PORT=6379
      - REDIS_QUEUE_SERVER_DB=0
      - MIXED_STORAGE_DETECTOR_STORAGE=tc_redis.storages.redis_storage
      - S3_USE_SIGV4=false
      - LOADER=tc_aws.loaders.s3_loader
      - TC_AWS_REGION=us-east-1
      - TC_AWS_LOADER_BUCKET=default
      - TC_AWS_ENDPOINT="http://s3:9000"
      - AWS_ACCESS_KEY_ID=minio
      - AWS_SECRET_ACCESS_KEY=minio123
    ports:
      - 8080:8000

  new:
    image: thumbor:dev
    build: thumbor/
    volumes:
      - thumbor-data:/data
      - thumbor-logs:/logs
      - ./thumbor/thumbor.conf:/usr/src/thumbor.conf
    command:
      - thumbor
      - -c
      - /usr/src/thumbor.conf
    environment:
      - DETECTORS=['thumbor.detectors.queued_detector.queued_complete_detector']
      - STORAGE=thumbor.storages.mixed_storage
      - REDIS_STORAGE_SERVER_HOST=redis
      - REDIS_STORAGE_SERVER_PORT=6379
      - REDIS_STORAGE_SERVER_DB=0
      - REDIS_QUEUE_SERVER_HOST=redis
      - REDIS_QUEUE_SERVER_PORT=6379
      - REDIS_QUEUE_SERVER_DB=0
      - MIXED_STORAGE_DETECTOR_STORAGE=tc_redis.storages.redis_storage
      - S3_USE_SIGV4=false
      - LOADER=tc_aws.loaders.s3_loader
      - TC_AWS_REGION=us-east-1
      - TC_AWS_LOADER_BUCKET=default
      - TC_AWS_STORAGE_BUCKET=default
      - TC_AWS_ENDPOINT="http://s3:9000"
      - AWS_ACCESS_KEY_ID=minio
      - AWS_SECRET_ACCESS_KEY=minio123
    ports:
      - 9999:8888

  redis:
    image: redis:latest
    volumes:
      - redis-data:/data
```

Bellow, I post the results for both thumbor 6.3 and thumbor 7.0.7 with the new plugin.

Reply:
Thanks for the performance tests. 
We did load testing before so it was on "live" servers with a lot of simulated users, so not really comparable, but here a wait time of ~3/4secs looks quite slow.
Ideally we should do some load testing on a live server with AWS S3, and do a before (Thumbor 6) / current (Thumbor 7 with latest tc_aws without your PR) / after (Thumbor 7 + your PR) to check the improvements.